File: src/RefCommon.cpp
#include "../include/RefCommon.h"
#include "../include/GenoRecord.h"
#include "../include/VCFGeno.h"
#include "../include/VCFFamily.h"

using namespace std;


//////////////////// RefCommon ////////////////////

vector<pair<size_t, vector<int>>>
RefCommon::merge_records_core(const VCFGenoBase *phased_vcf,
								const VCFGenoBase *non_phased_vcf,
								const STRVEC& samples) {
	// fill with N/A
	const size_t	N = samples.size();
	const size_t	M = phased_vcf->size();
	vector<pair<size_t, vector<int>>>	genos(M);
	for(size_t i = 0; i < M; ++i) {
		const GenoRecord	*record = phased_vcf->get_record(i);
		const ll	pos = record->get_pos();
		const vector<int>	geno(N, Genotype::NA);
		genos[i] = make_pair(pos, geno);
	}
	
	// Assign genotypes from the phased VCF
	const auto	cs1 = phased_vcf->extract_columns(samples);
	const auto	cs2 = non_phased_vcf->extract_columns(samples);
	for(size_t j = 0; j < M; ++j) {
		const GenoRecord	*record1 = phased_vcf->get_record(j);
		const auto&	geno = record1->get_genos();
		for(size_t i = 0; i < cs1.size(); ++i) {
			const size_t	c = cs1[i];
			if(c != string::npos)
				genos[j].second[i] = geno[c];
		}
	}
	
	size_t	k = 0;
	size_t	l = 0;
	const size_t	L1 = phased_vcf->size();
	const size_t	L2 = non_phased_vcf->size();
	while(k < L1 && l < L2) {
		const GenoRecord	*record1 = phased_vcf->get_record(k);
		const GenoRecord	*record2 = non_phased_vcf->get_record(l);
		if(record1->get_pos() == record2->get_pos()) {
			const auto&	geno = record2->get_genos();
			for(size_t i = 0; i < cs2.size(); ++i) {
				const size_t	c = cs2[i];
				if(c != string::npos && cs1[i] == string::npos)
					genos[k].second[i] = geno[c];
			}
		}
		if(record1->get_pos() <= record2->get_pos()) {
			k += 1;
		}
		if(record1->get_pos() >= record2->get_pos()) {
			l += 1;
		}
	}
	
	return genos;
}

vector<GenoRecord *> RefCommon::merge_records(const VCFGenoBase *phased_vcf,
											const VCFGenoBase *non_phased_vcf,
											const STRVEC& samples) {
	const auto	genos = merge_records_core(phased_vcf, non_phased_vcf, samples);
	const size_t	M = genos.size();
	vector<GenoRecord *>	records(M);
	for(size_t i = 0; i < M; ++i) {
		records[i] = new GenoRecord(genos[i].first, genos[i].second);
	}
	return records;
}

vector<VCFFamilyRecord *> RefCommon::merge_family_records(
											const VCFGenoBase *phased_vcf,
											const VCFGenoBase *non_phased_vcf,
											const STRVEC& samples) {
	const auto	genos = merge_records_core(phased_vcf, non_phased_vcf, samples);
	const size_t	M = genos.size();
	vector<VCFFamilyRecord *>	records(M);
	for(size_t i = 0; i < M; ++i) {
		records[i] = new VCFFamilyRecord(genos[i].first, genos[i].second);
	}
	return records;
}

vector<GenoRecord *> RefCommon::expand_records(const VCFGeno *vcf,
												const VCFGeno *phased_vcf) {
	if(vcf->size() == 0)
		return vector<GenoRecord *>();
	
	// fill with N/A
	const size_t	N = vcf->get_samples().size();
	const size_t	M = phased_vcf->size();
	vector<GenoRecord *>	records(M);
	for(size_t i = 0; i < M; ++i) {
		const GenoRecord	*record = phased_vcf->get_record(i);
		const ll	pos = record->get_pos();
		const vector<int>	geno(N, Genotype::NA);
		records[i] = new GenoRecord(pos, geno);
	}
	
	size_t	k = 0;
	size_t	l = 0;
	const size_t	L1 = vcf->size();
	const size_t	L2 = phased_vcf->size();
	while(k < L1 && l < L2) {
		const GenoRecord	*record1 = vcf->get_record(k);
		GenoRecord	*record2 = records[l];
		if(record1->get_pos() == record2->get_pos()) {
			record2->copy_genotypes_from(record1);
		}
		if(record1->get_pos() <= record2->get_pos()) {
			k += 1;
		}
		if(record1->get_pos() >= record2->get_pos()) {
			l += 1;
		}
	}
	
	return records;
}

File: python/RefCommon.py
from __future__ import annotations

# coding: utf-8
# RefCommon.py

from VCF import VCFSmall
from GenoRecord import GenoRecord
from VCFGeno import VCFGenoBase, VCFGeno
from VCFFamily import VCFFamilyRecord
from KnownFamily import KnownFamily
from Genotype import Genotype
from typing import Optional


#################### RefCommon ####################

# phasedのrecordとnon-phasedのrecordをmergeする
# phasedにしかないpositionは./.で埋めて
# phasedにないpositionは捨てる
def merge_records_core(phased_vcf: VCFGenoBase,
						non_phased_vcf: VCFGenoBase,
						samples: list[str]) -> list[tuple[int, list[int]]]:
	# 全部N/Aで埋める
	N = len(samples)
	M = len(phased_vcf)
	poss = [ phased_vcf.get_record(i).pos for i in range(M) ]
	genos: list[tuple[int, list[int]]] = [ (pos, [Genotype.NA]*N)
														for pos in poss ]
	cs1 = phased_vcf.extract_columns(samples)
	cs2 = non_phased_vcf.extract_columns(samples)
	
	# phasedのサンプルのGenotypeを入れる
	for j in range(M):
		record1 = phased_vcf.get_record(j)
		for i, c in enumerate(cs1):
			if c != -1:
				genos[j][1][i] = record1.geno[c]
	
	k, l = 0, 0
	L1, L2 = len(phased_vcf), len(non_phased_vcf)
	while k < L1 and l < L2:
		record1 = phased_vcf.get_record(k)
		record2 = non_phased_vcf.get_record(l)
		if record1.pos == record2.pos:
			for i, c in enumerate(cs2):
				if c != -1 and cs1[i] == -1:
					genos[k][1][i] = record2.geno[c]
			k += 1
			l += 1
		elif record1.pos < record2.pos:
			k += 1
		else:
			l += 1
	
	return genos

def merge_records(phased_vcf: VCFGenoBase,
						non_phased_vcf: VCFGenoBase,
						samples: list[str]) -> list[GenoRecord]:
	genos = merge_records_core(phased_vcf, non_phased_vcf, samples)
	return [ GenoRecord(pos, geno) for pos, geno in genos ]

def merge_family_records(phased_vcf: VCFGenoBase,
						non_phased_vcf: VCFGenoBase,
						samples: list[str]) -> list[VCFFamilyRecord]:
	genos = merge_records_core(phased_vcf, non_phased_vcf, samples)
	return [ VCFFamilyRecord(pos, geno) for pos, geno in genos ]

# phased_vcfのポジションでvcfに無いところはN/Aにする
def expand_records(vcf: VCFGeno, phased_vcf: VCFGeno) -> list[GenoRecord]:
	if len(vcf) == 0:
		return []
	
	N = len(vcf.samples)
	M = len(phased_vcf)
	poss = [ phased_vcf.records[i].pos for i in range(M) ]
	records = [ GenoRecord(pos, [Genotype.NA]*N) for pos in poss ]
	
	k, l = 0, 0
	L1, L2 = len(vcf), len(records)
	while k < L1 and l < L2:
		record1 = vcf.records[k]
		record2 = records[l]
		if record1.pos == record2.pos:
			for i in range(N):
				record2.geno[i] = record1.geno[i]
			k += 1
			l += 1
		elif record1.pos < record2.pos:
			k += 1
		else:
			l += 1
	
	return records
